api-gateway/src/main/java/com/petclinic/bffapigateway/utils/Security/Filters/IsUserFilter.java

```diff
@@ -0,0 +1,157 @@
+package com.petclinic.bffapigateway.utils.Security.Filters;
+
+
+import com.petclinic.bffapigateway.exceptions.HandlerIsNullException;
+import com.petclinic.bffapigateway.exceptions.InvalidInputException;
+import com.petclinic.bffapigateway.exceptions.InvalidTokenException;
+import com.petclinic.bffapigateway.utils.Security.Annotations.IsUserSpecific;
+import com.petclinic.bffapigateway.utils.Security.Annotations.SecuredEndpoint;
+import com.petclinic.bffapigateway.utils.Security.Variables.Roles;
+import lombok.Generated;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.java.Log;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.annotation.Order;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Component;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.reactive.HandlerMapping;
+import org.springframework.web.reactive.result.method.annotation.RequestMappingHandlerMapping;
+import org.springframework.web.server.ServerWebExchange;
+import org.springframework.web.server.WebFilter;
+import org.springframework.web.server.WebFilterChain;
+import reactor.core.publisher.Mono;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+@Order(3)
+@Generated
+public class IsUserFilter implements WebFilter {
+
+    private final JwtTokenUtil jwtTokenUtil;
+
+    private final RequestMappingHandlerMapping requestMappingHandlerMapping;
+
+
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
+
+        log.debug("IsUserFilter");
+        if (exchange.getAttribute("whitelisted") != null && exchange.getAttribute("whitelisted") instanceof Boolean) {
+            if((boolean) exchange.getAttribute("whitelisted")) {
+                return chain.filter(exchange);
+            }
+        }
+
+        HandlerMethod handler;
+        try {
+            handler = (HandlerMethod) requestMappingHandlerMapping.getHandler(exchange).toFuture().get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new HandlerIsNullException(e.getMessage());
+        }
+
+        if (handler == null) {
+            throw new HandlerIsNullException("Handler is null, check if the endpoint is valid");
+        }
+
+        if (handler.getMethod().getAnnotation(IsUserSpecific.class) == null) {
+            return chain.filter(exchange);
+        }
+
+
+        if (Arrays.stream(handler.getMethod().getAnnotation(IsUserSpecific.class).bypassRoles()).anyMatch(role -> role == Roles.ALL)
+        || Arrays.stream(handler.getMethod().getAnnotation(IsUserSpecific.class).bypassRoles()).anyMatch(role -> role == Roles.ANONYMOUS)) {
```

Comment:
Can you explain the filtering thing, please? This code:   if (exchange.getAttribute("whitelisted") != null && exchange.getAttribute("whitelisted") instanceof Boolean) {
            if((boolean) exchange.getAttribute("whitelisted")) {
                return chain.filter(exchange);
            }
        } got me particularly confused.

Reply:
The previous filters do a check if the request is anonymous, if it is, we allow everyone to accept it. To avoid repeating the check I added an attribute to the request (exchange) that would tell my future filters that the request has been white listed, it is also found in the roles filter.